Tests/AppTests/__Snapshots__/PackageCollectionTests/test_sign_collection.1.json

```diff
@@ -0,0 +1,59 @@
xQzZ3VmhRWjQ4SXRDRDN5NndzSUc5d3RqOEJNSXkzUTg4UG5UM3pLMGtvR3NqK3pyVzVEdGxlSE5iTFBiVTZyZlFQRGdDU0M3RWhGaTUwMVR3TjIySVdxNk54a2tkVFZjR3ZMMEd6K1B2amNNM21vMHhGZmg5TWExQ1dRWW5FZEdJTEVJTkJoek9LZ2JFd1dPeGFCREtNYUxPUEhkNWxjLzluWG1XOFNkaDJuek1VWmFGM2xNa3RBZ01CQUFHamdnRjZNSUlCZGpBT0JnTlZIUThCQWY4RUJBTUNBUVl3RHdZRFZSMFRBUUgvQkFVd0F3RUIvekFkQmdOVkhRNEVGZ1FVSzlCcFI1UjJDZjcwYTQwdVFLYjNSMDEvQ0Y0d0h3WURWUjBqQkJnd0ZvQVVLOUJwUjVSMkNmNzBhNDB1UUtiM1IwMS9DRjR3Z2dFUkJnTlZIU0FFZ2dFSU1JSUJCRENDQVFBR0NTcUdTSWIzWTJRRkFUQ0I4akFxQmdnckJnRUZCUWNDQVJZZWFIUjBjSE02THk5M2QzY3VZWEJ3YkdVdVkyOXRMMkZ3Y0d4bFkyRXZNSUhEQmdnckJnRUZCUWNDQWpDQnRocUJzMUpsYkdsaGJtTmxJRzl1SUhSb2FYTWdZMlZ5ZEdsbWFXTmhkR1VnWW5rZ1lXNTVJSEJoY25SNUlHRnpjM1Z0WlhNZ1lXTmpaWEIwWVc1alpTQnZaaUIwYUdVZ2RHaGxiaUJoY0hCc2FXTmhZbXhsSUhOMFlXNWtZWEprSUhSbGNtMXpJR0Z1WkNCamIyNWthWFJwYjI1eklHOW1JSFZ6WlN3Z1kyVnlkR2xtYVdOaGRHVWdjRzlzYVdONUlHRnVaQ0JqWlhKMGFXWnBZMkYwYVc5dUlIQnlZV04wYVdObElITjBZWFJsYldWdWRITXVNQTBHQ1NxR1NJYjNEUUVCQlFVQUE0SUJBUUJjTnBsTUxYaTM3WXliM1BOM20vSjIwbmN3VDhFZmhZT0ZHNWs5UnpmeXFadEFqaXpVc1pBUzJMNzBjNXZ1MG1RUHkzbFBOTmlpUHZsNC8ydklCK3g5T1lPTFV5RFRPTVN4djVwUENtdi9LL3hacHdVSmZCZEFWaEVlZE5PM2l5TTdSNlBWYnlUaTY5RzNjTjhQUmVFbnl2RnRlTzNudFJjWHFOeCtJalhLSmRYWkQ5WnIxS0lrSXhIM29heVBjNEZneGh0YkNTK1NzdmhFU1BCZ09KNFY5VDBtWnlDS00ycjNEWUxQM3V1akwvbFRhbHRrd0dNemQvYzZCeXhXNjlvUElRN2F1bk1aVDdYWk5uL0JoMVhacDVtNU1rTDcyTlZ4bm42aFVyY2J2Wk5DSkJJcXh3OGR0azJjWG1QSVM0QVhVS3FLMWRyay9OQUpCemV3ZFhVaCIKICBdCn0.ewogICJmb3JtYXRWZXJzaW9uIiA6ICIxLjAiLAogICJnZW5lcmF0ZWRBdCIgOiAiMTk3MC0wMS0wMVQwMDowMDowMFoiLAogICJrZXl3b3JkcyIgOiBbCgogIF0sCiAgIm5hbWUiIDogIkNvbGxlY3Rpb24iLAogICJvdmVydmlldyIgOiAiU29tZSBjb2xsZWN0aW9uIiwKICAicGFja2FnZXMiIDogWwogICAgewogICAgICAidXJsIiA6ICJ1cmwiLAogICAgICAidmVyc2lvbnMiIDogWwogICAgICAgIHsKICAgICAgICAgICJjcmVhdGVkQXQiIDogIjE5NzAtMDEtMDFUMDA6MDA6MDBaIiwKICAgICAgICAgICJkZWZhdWx0VG9vbHNWZXJzaW9uIiA6ICI1LjUiLAogICAgICAgICAgIm1hbmlmZXN0cyIgOiB7CiAgICAgICAgICAgICI1LjUiIDogewogICAgICAgICAgICAgICJwYWNrYWdlTmFtZSIgOiAiZm9vIiwKICAgICAgICAgICAgICAicHJvZHVjdHMiIDogWwogICAgICAgICAgICAgICAgewogICAgICAgICAgICAgICAgICAibmFtZSIgOiAicCIsCiAgICAgICAgICAgICAgICAgICJ0YXJnZXRzIiA6IFsKICAgICAgICAgICAgICAgICAgICAidCIKICAgICAgICAgICAgICAgICAgXSwKICAgICAgICAgICAgICAgICAgInR5cGUiIDogewogICAgICAgICAgICAgICAgICAgICJleGVjdXRhYmxlIiA6IG51bGwKICAgICAgICAgICAgICAgICAgfQogICAgICAgICAgICAgICAgfQogICAgICAgICAgICAgIF0sCiAgICAgICAgICAgICAgInRhcmdldHMiIDogWwogICAgICAgICAgICAgICAgewogICAgICAgICAgICAgICAgICAibmFtZSIgOiAidCIKICAgICAgICAgICAgICAgIH0KICAgICAgICAgICAgICBdLAogICAgICAgICAgICAgICJ0b29sc1ZlcnNpb24iIDogIjUuNSIKICAgICAgICAgICAgfQogICAgICAgICAgfSwKICAgICAgICAgICJ2ZXJzaW9uIiA6ICIxLjIuMyIKICAgICAgICB9CiAgICAgIF0KICAgIH0KICBdCn0.isrE2T3a5C6FFqx3Xtqo_We6ftiLolUJisToW1jsn9YPXDZQEdAfXcovtmN8YY_uIrNXt-mEe7gpNkOSSKW98NHOnXeNnF3eFA5Wu6Av-l7_BBUvvAheGY1ab1HOIa6ZrLez9L6BLPckr1vMegjyjwLoYnWXHaQ7qAXTTuzArf31vP5wekoLuoS5gfRWkD6OmFcpbk6zHum-x4xLTilHuqOwFL7vPD-kaCsVdI1ayc7mySialqswwo9ius0q9A4kpQMNUDQ3pRYVZjvaeDr9AhqPpxE-uTTmOC2FOtn3mPnImkCT9FiBgd5xX4a--QdhQQ84vyXQEQBiN417lS7TEQ"
```

Comment:
Something I found out while reading the signing code: This is a [JWT encoded payload + signature](https://github.com/apple/swift-package-manager/blob/signer-default-observability-scope/Sources/PackageCollectionsSigning/Signing/Signature.swift#L66):

```swift
    static func generate<Payload>(for payload: Payload,
                                  with header: Header,
                                  using signer: MessageSigner,
                                  jsonEncoder: JSONEncoder) throws -> Data where Payload: Encodable {
        let headerData = try jsonEncoder.encode(header)
        let encodedHeader = headerData.base64URLEncodedBytes()

        let payloadData = try jsonEncoder.encode(payload)
        let encodedPayload = payloadData.base64URLEncodedBytes()

        // https://www.rfc-editor.org/rfc/rfc7515.html#section-5.1
        // Signing input: BASE64URL(header) + '.' + BASE64URL(payload)
        let signatureData = try signer.sign(message: encodedHeader + .period + encodedPayload)
        let encodedSignature = signatureData.base64URLEncodedBytes()

        // Result: header.payload.signature
        let bytes = encodedHeader
            + .period
            + encodedPayload
            + .period
            + encodedSignature
        return bytes
    }
```

which means that the whole collection is contained as a base64 encoded duplicate within the signature. For large collections that's going to create unnecessarily big collections.

I suppose the problems are

- if you don't include the plain text JSON the data is useless unless you have infra to decode the signature
- if you don't include the payload in the signature you rely on the json being passed in exactly as is for verification

On the other hand: I'm sure Xcode has to throw away the plain text JSON anyway and only use the encoded one, right? Otherwise it's trivial to pass in mismatched signed/compromised content.

And it seems like it would be very tempting for consumers to treat the signature as authoritative and then use the plain JSON for simplicity. So arguably it might be better not to ship it in the first place.

I feel like we should discuss this with the SPM team.

Reply:
I think discussing it with them is a good idea, but I think the first question should be to check that the encoded version is used in that way. I thought the signature was checked as a confirmation of the validity of the JSON data, not as something that would be decoded and used.

The other obvious disadvantage of *only* including the encoded version is that it presents a significant barrier to people just wanting to experiment with using collection data. If we decided to only include the signature, we'd immediately be trying to give a URL parameter or different URL location for an unsigned version of the data.